scripts/scratch_ingest/expression_writer.py

```diff
@@ -0,0 +1,435 @@
+            while current_pos < end_pos:
+                line = matrix_file.readline()
+                gene_idx = int(line.split()[0])
+                gene_name = genes[gene_idx - 1]
+                fragment_path = f"{data_dir}/gene_entries/{gene_name}__entries.txt"
+                with open(fragment_path, 'a+') as file:
+                    file.write(line)
+                current_pos += len(line)
+
+    def process_sparse_data_fragments(self, barcodes, cluster_cells, cluster_name, data_dir):
+        """
+        Find and process all generated single-gene sparse data fragments
+
+        :param barcodes: (List) list of cell barcodes
+        :param cluster_cells: (List) list of cells from cluster file
+        :param cluster_name: (String) name of cluster object
+        :param data_dir: (String) name out output dir
+        """
+        fragments = os.listdir(f"{data_dir}/gene_entries")
+        print(f"subdivision complete, processing {len(fragments)} fragments")
+        pool = multiprocessing.Pool(self.num_cores)
+        processor = partial(self.process_fragment,
+                            barcodes=barcodes, cluster_cells=cluster_cells,
+                            cluster_name=cluster_name, data_dir=data_dir)
+        pool.map(processor, fragments)
+
+    def process_fragment(self, fragment_name, barcodes, cluster_cells, cluster_name, data_dir):
+        """
+        Process a single-gene sparse matrix fragment and write expression array
+
+        :param fragment_name: (String) name of gene-level fragment file
+        :param barcodes: (List) list of cell barcodes
+        :param cluster_cells: (List) list of cells from cluster file
+        :param cluster_name: (String) name of cluster object
+        :param data_dir: (String) name out output dir
+        """
+        gene_name = fragment_name.split('__')[0]
+        full_path = f"{data_dir}/gene_entries/{fragment_name}"
+        observed_cells = []
+        exp_vals = []
+        for line in ExpressionWriter.open_file(full_path):
+            gene_idx, barcode_idx, exp_val = self.extract_sparse_line(line)
+            observed_cells.append(barcodes[barcode_idx - 1])
+            exp_vals.append(exp_val)
+        filtered_expression = ExpressionWriter.filter_expression_for_cluster(
+            cluster_cells, observed_cells, exp_vals
+        )
+        ExpressionWriter.write_gene_scores(gene_name, cluster_name, filtered_expression, data_dir)
+
+    def extract_sparse_line(self, line):
+        """
+        Process a single line from a sparse matrix and extract values as integers
+
+        :param line: (String) single line from matrix file
+        :return: (List): values as integers
+        """
+        gene_idx, barcode_idx, raw_exp = line.rstrip().split(' ')
+        return [int(gene_idx), int(barcode_idx), round(float(raw_exp), self.precision)]
+
+    def write_empty_sparse_genes(self, genes, num_cluster_cells, cluster_name, data_dir):
```

Comment:
Optimization idea maybe worth a TODO: 

Write a list of gene symbols that lack expression in a special file, then background-read that into the frontend upon loading a cluster, and display a “0” gene expression scatter plot for all of them.  

Given ~4k of 28k genes were empty in our pilot, that would save a sizable amount of total data size and run time here, as well as time for end users and Image Pipeline.

Reply:
I think that's definitely worth ticketing, but for now it will keep the code on your end cleaner.

Reply:
I just opened SCP-4683: "Optimize scatter data speed/cost 14% via unexpressed genes".